xchange-stream-kucoin/src/main/java/info/bitrich/xchangestream/kucoin/KucoinStreamingMarketDataService.java

```diff
@@ -0,0 +1,187 @@
+  }
+
+  private Observable<KucoinOrderBookEventData> rawOrderBookUpdates(CurrencyPair currencyPair) {
+    String channelName = "/market/level2:" + KucoinAdapters.adaptCurrencyPair(currencyPair);
+
+    return service
+            .subscribeChannel(channelName)
+            .doOnError(ex -> logger.warn("encountered error while subscribing to channel " + channelName, ex))
+            .map(it -> mapper.treeToValue(it, KucoinOrderBookEvent.class))
+            .map(e -> e.data);
+  }
+
+
+  private Observable<OrderBook> createOrderBookObservable(CurrencyPair currencyPair) {
+    // 1. Open a stream
+    // 2. Buffer the events you receive from the stream.
+    OrderbookSubscription subscription =
+            new OrderbookSubscription(orderBookRawUpdatesSubscriptions.get(currencyPair));
+
+    return subscription
+            .stream
+            // 3. Get a depth snapshot
+            // (we do this if we don't already have one or we've invalidated a previous one)
+            .doOnNext(transaction -> subscription.initSnapshotIfInvalid(currencyPair))
+
+            .doOnError(ex -> logger.warn("encountered error while processing order book event", ex))
+
+            // If we failed, don't return anything. Just keep trying until it works
+            .filter(transaction -> subscription.snapshotLastUpdateId.get() > 0L)
+
+            // 4. Drop any event where u is <= lastUpdateId in the snapshot
+            .filter(depth -> depth.sequenceEnd > subscription.snapshotLastUpdateId.get())
+
+            // 5. The first processed should have U <= lastUpdateId+1 AND u >= lastUpdateId+1, and
+            // subsequent events would
+            // normally have u == lastUpdateId + 1 which is stricter version of the above - let's be
+            // more relaxed
+            // each update has absolute numbers so even if there's an overlap it does no harm
+            .filter(
+                    depth -> {
+                      long lastUpdateId = subscription.lastUpdateId.get();
+                      boolean result;
+                      if (lastUpdateId == 0L) {
+                        result = true;
+                      } else {
+                        result =
+                                depth.sequenceStart <= lastUpdateId + 1
+                                        && depth.sequenceEnd >= lastUpdateId + 1;
+                      }
+                      if (result) {
+                        subscription.lastUpdateId.set(depth.sequenceEnd);
+                      } else {
+                        // If not, we re-sync
+                        logger.info(
+                                "Orderbook snapshot for {} out of date (last={}, U={}, u={}). This is normal. Re-syncing.",
+                                currencyPair,
+                                lastUpdateId,
+                                depth.sequenceStart,
+                                depth.sequenceEnd);
+                        subscription.invalidateSnapshot();
```

Comment:
should this not return false?

Reply:
I've copied this code from binance where it exists for a few years